File: template/php/system.html.php
<?php
// system.html.php
// to run: php system.html.php

function run_script($script_name) {
	$output = '';
	$status = 'success';

	if (substr($script_name, -3) === '.js') {
		exec("node $script_name 2>&1", $output, $return_var);
	} elseif (substr($script_name, -4) === '.php') {
		exec("php $script_name 2>&1", $output, $return_var);
	} elseif (substr($script_name, -3) === '.py') {
		exec("python $script_name 2>&1", $output, $return_var);
	} elseif (substr($script_name, -3) === '.sh') {
		exec("bash $script_name 2>&1", $output, $return_var);
	} else {
		return array("Unsupported file type: $script_name", "failure");
	}

	$output = implode("\n", $output);
	if ($return_var !== 0) {
		$status = 'failure';
		$output = "Error running $script_name: $output";
	}

	return array($output, $status);
}

function read_scripts_file($filename) {
	$scripts = array();
	$lines = file($filename, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);
	foreach ($lines as $line) {
		$script = trim($line);
		if ($script) {
			$function = explode('_', $script)[1];
			if (!isset($scripts[$function])) {
				$scripts[$function] = array();
			}
			$scripts[$function][] = $script;
		}
	}
	return $scripts;
}

function generate_html() {
	$scripts = read_scripts_file('scripts.txt');

	$summary_data = array();
	$detailed_data = array();

	foreach ($scripts as $function => $script_list) {
		$function_summary = array("function" => $function);
		$function_details = array("function" => $function, "scripts" => array());

		foreach (array('.js', '.php', '.py', '.sh') as $ext) {
			$script = null;
			foreach ($script_list as $s) {
				if (substr($s, -strlen($ext)) === $ext) {
					$script = $s;
					break;
				}
			}

			if ($script) {
				echo "Running $script...\n";
				list($output, $status) = run_script($script);

				if (!$output) {
					$summary = "No output";
					$status = "n/a";
				} elseif (stripos($output, "error") !== false) {
					$status = "failed";
					$summary = "failed";
				} else {
					$status = "success";
					$summary = "success";
				}

				$function_summary[substr($ext, 1)] = array("summary" => $summary, "status" => $status);
				$function_details["scripts"][] = array(
					"name" => $script,
					"summary" => $summary,
					"status" => $status,
					"details" => $output
				);
			} else {
				$function_summary[substr($ext, 1)] = array("summary" => "N/A", "status" => "n/a");
			}
		}

		$summary_data[] = $function_summary;
		$detailed_data[] = $function_details;
	}

	$template = file_get_contents('./template/html/system_template.html');
	$css_content = file_get_contents('./template/css/styles.css');

	$html_content = str_replace(
		array('{{ css_content }}', '{{ title }}', '{{ current_time }}'),
		array($css_content, 'Thimble System Report', date('Y-m-d H:i:s')),
		$template
	);

	// Replace summary_data and detailed_data placeholders
	$summary_html = '';
	foreach ($summary_data as $item) {
		$summary_html .= "<tr>
			<td>{$item['function']}</td>
			<td class=\"{$item['js']['status']}\">{$item['js']['summary']}</td>
			<td class=\"{$item['php']['status']}\">{$item['php']['summary']}</td>
			<td class=\"{$item['py']['status']}\">{$item['py']['summary']}</td>
			<td class=\"{$item['sh']['status']}\">{$item['sh']['summary']}</td>
		</tr>";
	}
	$html_content = str_replace('{{ summary_data }}', $summary_html, $html_content);

	$detailed_html = '';
	foreach ($detailed_data as $function) {
		$detailed_html .= "<div class=\"function-group\">
			<h3>{$function['function']}</h3>";
		foreach ($function['scripts'] as $script) {
			$detailed_html .= "<div class=\"script-details\">
				<h4>{$script['name']}</h4>
				<p class=\"summary {$script['status']}\">Summary: {$script['summary']}</p>
				<div class=\"details\">
					<h5>Details:</h5>
					<pre>{$script['details']}</pre>
				</div>
			</div>";
		}
		$detailed_html .= "</div>";
	}
	$html_content = str_replace('{{ detailed_data }}', $detailed_html, $html_content);

	file_put_contents('system.html', $html_content);
}

$start_time = microtime(true);
echo "Initiating system.html generation...\n";
generate_html();
$end_time = microtime(true);
$execution_time = $end_time - $start_time;
echo "system.html has been generated successfully. Time elapsed: " . number_format($execution_time, 2) . " seconds.\n";

File: template/python3/system.html.py
# system.html.py
# to run: python system.html.py

import os
import subprocess
import time
import asyncio
import aiofiles
from jinja2 import Environment, FileSystemLoader

async def run_script(script_name):
	try:
		if script_name.endswith('.js'):
			proc = await asyncio.create_subprocess_exec('node', script_name, stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE)
		elif script_name.endswith('.php'):
			proc = await asyncio.create_subprocess_exec('php', script_name, stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE)
		elif script_name.endswith('.py'):
			proc = await asyncio.create_subprocess_exec('python', script_name, stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE)
		elif script_name.endswith('.sh'):
			proc = await asyncio.create_subprocess_exec('bash', script_name, stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE)
		else:
			return f"Unsupported file type: {script_name}", "failure"

		stdout, stderr = await proc.communicate()
		if proc.returncode == 0:
			return stdout.decode().strip(), "success"
		else:
			return f"Error running {script_name}: {stderr.decode()}", "failure"
	except Exception as e:
		return f"Error running {script_name}: {str(e)}", "failure"

async def read_scripts_file(filename):
	scripts = {}
	async with aiofiles.open(filename, 'r') as f:
		lines = await f.readlines()
		for line in lines:
			script = line.strip()
			if script:
				function = script.split('_')[1]  # Assuming function name is the second part of the filename
				if function not in scripts:
					scripts[function] = []
				scripts[function].append(script)
	return scripts

async def generate_html():
	scripts = await read_scripts_file('scripts.txt')

	env = Environment(loader=FileSystemLoader('./template/html'))
	template = env.get_template('report_template.html')

	summary_data = []
	detailed_data = []

	for function, script_list in scripts.items():
		function_summary = {"function": function}
		function_details = {"function": function, "scripts": []}

		for ext in ['.js', '.php', '.py', '.sh']:
			script = next((s for s in script_list if s.endswith(ext)), None)
			if script:
				print(f"Running {script}...")
				output, status = await run_script(script)

				if not output:
					summary = "No output"
					status = "n/a"
				elif "error" in output.lower():
					status = "failed"
					summary = "failed"
				else:
					status = "success"
					summary = "success"

				function_summary[ext[1:]] = {"summary": summary, "status": status}
				function_details["scripts"].append({
					"name": script,
					"summary": summary,
					"status": status,
					"details": output
				})
			else:
				function_summary[ext[1:]] = {"summary": "N/A", "status": "n/a"}

		summary_data.append(function_summary)
		detailed_data.append(function_details)

	html_content = template.render(summary_data=summary_data, detailed_data=detailed_data)

	async with aiofiles.open('system.html', 'w') as f:
		await f.write(html_content)

async def main():
	start_time = time.time()
	print("Initiating system.html generation...")
	await generate_html()
	end_time = time.time()
	print(f"system.html has been generated successfully. Time elapsed: {end_time - start_time:.2f} seconds.")

if __name__ == "__main__":
	asyncio.run(main())

# end of system.html.py